### PR TITLE
Fixes the git lfs error by adding git lfs fetching

### PR DESCRIPTION
**Pull request by ernstluring**
This PR addresses issue #1462 and is a part of the solution that fixes the failing clone from a source with LFS files.
I have added a runtime check if the repo uses git lfs and a git lfs fetch call that fetches all LFS files, so it can be copied to the destination directory later on.

### CODE REVIEW

**Comment by jezdez** [comment]
@anaconda-issue-bot check

**Comment by baszalmstra** [comment]
Any idea why the pipelines fail? Is it because `git lfs` on CI is not the correct version?

**Comment by xhochy** [comment]
@jezdez Is there anything I can do to bring this over the finish line? I also run into the problem today and would like to get it fixed.

File: conda_build/source.py [hunk 1]
```diff
@@ -171,6 +171,17 @@ def unpack(source_dict, src_dir, cache_folder, recipe_path, croot, verbose=False
             shutil.move(os.path.join(tmpdir, f), os.path.join(src_dir, f))
 
 
+def repo_uses_git_lfs(git, dir):
+    lfs_list_output = check_output_env([git, 'lfs', 'ls-files', '--all'], cwd=dir)
+    return lfs_list_output and lfs_list_output.strip()
```
Comment (jezdez) [inline conda_build/source.py:176]: Not a fan that this doesn't catch the inevitable `CalledProcessError` exception inline, leading to the error output being swallowed in `git_mirror_checkout_recursive`.

```suggestion
def check_git_lfs(git, cwd):
    try:
        lfs_list_output = check_output_env([git, 'lfs', 'ls-files', '--all'], cwd=cwd)
        return lfs_list_output and lfs_list_output.strip()
    except CalledProcessError:
        return False
```

File: conda_build/source.py [hunk 2]
```diff
@@ -171,6 +171,17 @@ def unpack(source_dict, src_dir, cache_folder, recipe_path, croot, verbose=False
...
+
+
+def git_lfs_fetch(git, dir, stdout, stderr):
+    lfs_version = check_output_env([git, 'lfs', 'version'], cwd=dir)
+    print(lfs_version)
+    check_call_env([git, 'lfs', 'fetch', 'origin', '--all'], cwd=dir, stdout=stdout, stderr=stderr)
```
Comment (jezdez) [inline conda_build/source.py:182]:
```suggestion
def git_lfs_fetch(git, cwd, stdout, stderr):
    lfs_version = check_output_env([git, 'lfs', 'version'], cwd=cwd)
    log.info(lfs_version)
    check_call_env([git, 'lfs', 'fetch', 'origin', '--all'], cwd=cwd, stdout=stdout, stderr=stderr)
```

**Review by jezdez — COMMENTED** [review]
Just a few nits, nothing serious.

**Comment by beeankha** [comment]
pre-commit.ci autofix

**Review by beeankha — DISMISSED** [review]

File: news/gh-4318-Fixes-git-lfs-error.rst [hunk 3]
```diff
@@ -0,0 +1,24 @@
+Enhancements:
+-------------
+
+* <news item>
+
+Bug fixes:
+----------
+
+* Fixes the failing git clone when source has LFS files
```
Comment (kenodegard) [inline news/gh-4318-Fixes-git-lfs-error.rst:9]:
```suggestion
* Fixes the failing `git clone` when source has LFS files. (#4318)
```

**Review by kenodegard — APPROVED** [review]

**Comment by cwharris** [comment]
This update has monetary implications for anyone using conda-build with GIT LFS. Using LFS with GitHub, for instance, costs money. If someone is using conda-build for each commit but isn't already using LFS as part of their checks (build, test, etc), this update is going to start downloading LFS files for those users.

This feature should be opt-in by default.

**Comment by jakirkham** [comment]
Would it make sense to convert this request to an issue?

**Comment by h-vetinari** [comment]
I'm still seeing an error with conda-build 3.24.0 related to a `git_url:` source. I left a comment in https://github.com/conda/conda-build/issues/1462, but happy to open a new issue.

**Comment by jakirkham** [comment]
Yeah a new issue sounds like a good idea (just to make sure that doesn't get lost in a closed one)

**Comment by h-vetinari** [comment]
> Yeah a new issue sounds like a good idea (just to make sure that doesn't get lost in a closed one)

#4882

**Pull request merged**